### PR TITLE
[FLINK-11334][core] Migrate EnumValueSerializer to use new serialization compatibility abstractions

### PR DESCRIPTION
**Pull request by aljoscha**
This is an updated version of #7734 

It has the following changes on top:
 - Don't use IntSerializer in EnumValueSerializer
 - Remove old deserialization logic from ScalaEnumSerializerSnapshot
 - Create proper restore serializer in ScalaEnumSerializerSnapshot
 - Fix migration check in ScalaEnumSerializerSnapshot

The individual commits have explanations for why the change is necessary.

cc @klion26

### CODE REVIEW

File: flink-scala/src/main/scala/org/apache/flink/api/scala/typeutils/ScalaEnumSerializerSnapshot.scala [hunk 1]
```diff
@@ -73,7 +73,8 @@ class ScalaEnumSerializerSnapshot[E <: Enumeration]
   }
 
   override def restoreSerializer(): TypeSerializer[E#Value] = {
-    enumClass.newInstance().asInstanceOf[TypeSerializer[E#Value]]
+    val enumObject = enumClass.getField("MODULE$").get(null).asInstanceOf[E]
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-scala/src/main/scala/org/apache/flink/api/scala/typeutils/ScalaEnumSerializerSnapshot.scala]: I'm adding it there. Where else would you add it?

[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-scala/src/main/scala/org/apache/flink/api/scala/typeutils/ScalaEnumSerializerSnapshot.scala]: Good point, btw! 😃

[the remark this answers is not in the document]
Comment (klion26) [reply flink-scala/src/main/scala/org/apache/flink/api/scala/typeutils/ScalaEnumSerializerSnapshot.scala]: I just want to add this check in function `restoreStoreSerializer`, `resolveSchemaCompatibility` and `writeSnapshot`. sorry for forgot adding it in Scala code~

**Comment by aljoscha** [comment]
Merged. Thanks again @klion26 for your contribution!

**Pull request closed without merging**